### PR TITLE
Low mach number correction for HLLC Riemann solver

### PR DESCRIPTION
**Pull request by lee-hyeoksu**
## Description

The HLLC Riemann solver is known to have an accuracy issue at the low Mach number limit. The new feature addresses this issue using two different methods (Chen et al. 2022 and Thornber et al. 2008).

### Type of change
- [x] New feature (non-breaking change which adds functionality)

### Scope

- [x] This PR comprises a set of related changes with a common goal

If you cannot check the above box, please split your PR into multiple PRs that each have a common goal.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. 
Provide instructions so we can reproduce. 
Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:

* What computers and compilers did you use to test this:

## Checklist

- [x] I have added comments for the new code
- [x] I added Doxygen docstrings to the new code
- [x] I have made corresponding changes to the documentation (`docs/`)
- [x] I have added regression tests to the test suite so that people can verify in the future that the feature is behaving as expected
- [x] I have added example cases in `examples/` that demonstrate my new feature performing as expected.
They run to completion and demonstrate "interesting physics"
- [x] I ran `./mfc.sh format` before committing my code
- [ ] New and existing tests pass locally with my changes, including with GPU capability enabled (both NVIDIA hardware with NVHPC compilers and AMD hardware with CRAY compilers) and disabled
- [x] This PR does not introduce any repeated code (it follows the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle)
- [x] I cannot think of a way to condense this code and reduce any introduced additional line count

### If your code changes any code source files (anything in `src/simulation`)

To make sure the code is performing as expected on GPU devices, I have:
- [ ] Checked that the code compiles using NVHPC compilers
- [ ] Checked that the code compiles using CRAY compilers
- [ ] Ran the code on either V100, A100, or H100 GPUs and ensured the new feature performed as expected (the GPU results match the CPU results)
- [ ] Ran the code on MI200+ GPUs and ensure the new features performed as expected (the GPU results match the CPU results)
- [ ] Enclosed the new feature via `nvtx` ranges so that they can be identified in profiles
- [ ] Ran a Nsight Systems profile using `./mfc.sh run XXXX --gpu -t simulation --nsys`, and have attached the output file (`.nsys-rep`) and plain text results to this PR
- [ ] Ran an Omniperf profile using `./mfc.sh run XXXX --gpu -t simulation --omniperf`, and have attached the output file and plain text results to this PR.
- [ ] Ran my code using various numbers of different GPUs (1, 2, and 8, for example) in parallel and made sure that the results scale similarly to what happens if you run without the new code/feature

### CODE REVIEW

File: src/simulation/m_riemann_solvers.fpp [hunk 1]
```diff
@@ -1948,16 +1989,16 @@ contains
...
-                                                          s_M*(xi_L*(dir_flg(dir_idx(1))*s_S + &
-                                                                     (1d0 - dir_flg(dir_idx(1)))* &
-                                                                     vel_L(dir_idx(1))) - vel_L(dir_idx(1))))) &
-                                            - xi_P*(rho_R*(vel_R(dir_idx(1))* &
-                                                           vel_R(dir_idx(1)) + &
-                                                           s_P*(xi_R*(dir_flg(dir_idx(1))*s_S + &
-                                                                      (1d0 - dir_flg(dir_idx(1)))* &
-                                                                      vel_R(dir_idx(1))) - vel_R(dir_idx(1)))))
+                                            -xi_M*(rho_L*(vel_L(idx1)* &
+                                                          vel_L(idx1) + &
```
Comment (sbryngelson) [inline src/simulation/m_riemann_solvers.fpp:1975]: what motivated the change from dir_idx(1) to idx1?

File: src/simulation/m_riemann_solvers.fpp [hunk 2]
```diff
@@ -1387,6 +1396,15 @@ contains
                                 xi_M = (5d-1 + sign(5d-1, s_S))
                                 xi_P = (5d-1 - sign(5d-1, s_S))
 
+                                if (low_Mach == 1) then
+                                    pcorr = rho_L*rho_R* &
```
Comment (sbryngelson) [inline src/simulation/m_riemann_solvers.fpp]: these `if(low_mach = 1` (or `2`) routines seem to be repeated for all of the types of Riemann solver. they should be turned into their own functions or subroutines.

File: src/simulation/m_riemann_solvers.fpp [hunk 3]
```diff
@@ -1764,20 +1785,28 @@ contains
                                     end do
                                 end if
 
+                                if (low_Mach == 2) then
+                                    zcoef = min(1d0, max(vel_L_rms**0.5d0/c_L, vel_R_rms**0.5d0/c_R))
```
Comment (sbryngelson) [inline src/simulation/m_riemann_solvers.fpp]: these `if(low_mach = 1` (or `2`) routines seem to be repeated for all of the types of Riemann solver. they should be turned into their own functions or subroutines.

File: src/simulation/m_riemann_solvers.fpp [hunk 4]
```diff
@@ -1802,21 +1831,30 @@ contains
...
+                                xi_R = (s_R - vel_R(idx1))/(s_R - s_S)
 
                                 ! goes with numerical velocity in x/y/z directions
                                 ! xi_P/M = 0.5 +/m sgn(0.5,s_star)
                                 xi_M = (5d-1 + sign(5d-1, s_S))
                                 xi_P = (5d-1 - sign(5d-1, s_S))
 
+                                if (low_Mach == 1) then
+                                    pcorr = rho_L*rho_R* &
+                                            (s_L - vel_L(idx1))*(s_R - vel_R(idx1))*(vel_R(idx1) - vel_L(idx1))/ &
```
Comment (sbryngelson) [inline src/simulation/m_riemann_solvers.fpp]: these `if(low_mach = 1` (or `2`) routines seem to be repeated for all of the types of Riemann solver. they should be turned into their own functions or subroutines.

File: src/simulation/m_riemann_solvers.fpp [hunk 5]
```diff
@@ -2108,6 +2149,14 @@ contains
                                     end do
                                 end if
 
+                                if (low_Mach == 2) then
+                                    zcoef = min(1d0, max(vel_L_rms**0.5d0/c_L, vel_R_rms**0.5d0/c_R))
+                                    qL_prim_rs${XYZ}$_vf(j, k, l, contxe + idx1) = 5d-1*((vel_L(idx1) + vel_R(idx1)) + zcoef*(vel_L(idx1) - vel_R(idx1)))
+                                    qR_prim_rs${XYZ}$_vf(j + 1, k, l, contxe + idx1) = 5d-1*((vel_L(idx1) + vel_R(idx1)) + zcoef*(vel_R(idx1) - vel_L(idx1)))
```
Comment (sbryngelson) [inline src/simulation/m_riemann_solvers.fpp]: these `if(low_mach = 1` (or `2`) routines seem to be repeated for all of the types of Riemann solver. they should be turned into their own functions or subroutines.

File: src/simulation/m_riemann_solvers.fpp [hunk 6]
```diff
@@ -2155,6 +2204,15 @@ contains
                                 xi_M = (5d-1 + sign(5d-1, s_S))
                                 xi_P = (5d-1 - sign(5d-1, s_S))
 
+                                if (low_Mach == 1) then
+                                    pcorr = rho_L*rho_R* &
+                                            (s_L - vel_L(idx1))*(s_R - vel_R(idx1))*(vel_R(idx1) - vel_L(idx1))/ &
+                                            (rho_R*(s_R - vel_R(idx1)) - rho_L*(s_L - vel_L(idx1)))* &
```
Comment (sbryngelson) [inline src/simulation/m_riemann_solvers.fpp]: these `if(low_mach = 1` (or `2`) routines seem to be repeated for all of the types of Riemann solver. they should be turned into their own functions or subroutines.

File: src/simulation/m_riemann_solvers.fpp [hunk 7]
```diff
@@ -1512,12 +1531,14 @@ contains
                     end do
 
                 elseif (model_eqns == 2 .and. bubbles) then
```
Comment (sbryngelson) [inline src/simulation/m_riemann_solvers.fpp:1524]: does this feature work with `bubbles == .TRUE.`? If not, then remove the modified code. IF it does, then there should be CI tests for it

File: src/simulation/m_riemann_solvers.fpp (hunk 1 above)
Comment (hyeoksu-lee) [reply src/simulation/m_riemann_solvers.fpp:1975]: For consistency with `else` part. `else if (model_eqns == 2 .and. bubbles)` code is very similar to `else` code, so I wanted to make them have same things. I thought `idx1` is more concise than `dir_idx(1)`.

File: src/simulation/m_riemann_solvers.fpp (hunk 2 above)
Comment (hyeoksu-lee) [reply src/simulation/m_riemann_solvers.fpp]: I move them into a subroutine

File: src/simulation/m_riemann_solvers.fpp (hunk 1 above)
Comment (hyeoksu-lee) [reply src/simulation/m_riemann_solvers.fpp:1975]: There are more places using `dir_idx(1)` and `dir_idx(i)` in the code. If you agree, I will change them to `idx1` and `idxi`. Or, if you prefer `dir_idx(1)` and `dir_idx(i)`, I will change `idx1` and `idxi` to them. I just want to make them consistent.

**Pull request closed without merging**